Changgyo

Dunkkkk opened the pull request:
과제를 완료했습니다. 부족한 부분 강력한 충고를 감사드립니당~

Dunkkkk committed:
Make titleView and basic ObservableObject for gamelogic

Make TitleView and Them, plus, minus button ( just appearence ) and make ObservableObject to convert EnvironmentObject ( this will excute basic game logic and store every properties)

Dunkkkk committed:
Add logic	that can be plus & minus card and  when user Shuffled, change the card

Add logic that can be plus & minus card ( without random shuffle ), and if user change theme, EnvironmentObject must change cardCount, CardTheme, Cardlist

Dunkkkk committed:
Change some variable name and Fix code convention

Chang som variable name in MemorizeCardGameManeger and Fix some code convention

gen-com commented on week1/Memorize/Memorize/MemorizeApp.swift:
```diff
@@ -0,0 +1,21 @@
+
+import SwiftUI
+
+
```
무의미한 공백 관리 해주세요.

Dunkkkk replied:
코드컨벤션 관련한 부분은 코드리뷰를 기반으로 전부 수정했습니다!

gen-com commented on week1/Memorize/Memorize/View/MainGameBoardView.swift:
```diff
@@ -0,0 +1,28 @@
+
+struct MainGameBoardView: View {
+    
+   
```
무의미한 공백 관리

gen-com commented on week1/Memorize/Memorize/View/TitleView.swift:
```diff
@@ -0,0 +1,21 @@
+struct TitleView: View {
+    var body: some View {
+        Text("Memorize!")
+            .font(.custom("", fixedSize: 45))
```
custom("", fixedSize: 45) 커스텀 폰트를 사용하면서, 폰트 이름을 비워둔 이유는 무엇인가요 ?
시스템 폰트를 사용하고 싶었던 거면 custom을 사용할 필요가 있나요 ?

Dunkkkk replied:
습관적인 사용이였던것 같습니다. systemFont로 수정완료했고, 다음부터 systemFont로 사용하겠습니다!

gen-com commented on week1/Memorize/Memorize/View/MemorizeButtonListVIew.swift:
```diff
@@ -0,0 +1,31 @@
+import SwiftUI
+
+struct MemorizeButtonListView: View {
+    
```
선언 후 공백을 둘 지 두지 않을 지 통일해 주세요.
struct MemorizeButtonListView: View {
@EnvironmentObject var gameManager: MemorizeCardGameManger

or 

struct MemorizeButtonListView: View {

@EnvironmentObject var gameManager: MemorizeCardGameManger

gen-com commented on week1/Memorize/Memorize/View/MemorizeButtonListVIew.swift:
```diff
@@ -0,0 +1,31 @@
+
+struct MemorizeButtonListView: View {
+    
+    @EnvironmentObject var gameManager: MemorizeCardGameManger
```
@EnvironmentObject 를 사용한 이유는 무엇인가요 ?

@ObservedObject와 어떤 차이가 있나요 ?

Dunkkkk replied:
@ObservedObject는 state처럼 값을 변경할수 있으면 외부 참조타입을 사용할 수 있는 property wrapper입니다.
여기서 Published로 선언 변수는 해당 변수를 observe하고 state가 변하면 해당 관련되 view를 reload할수 있습니다.
여기서 @EnvironmentObject와 차이는 Binding을 통해 직접 내려야 한다. 상위뷰에 설정만 해두면 어디서든 사용할수 있다는 점입니다.
저는 button과 buttonlist를 만들었는데, 굳이 buttonlist에선 바인딩하지 않아도 되는 경우가 있기때문에 중간에 Binding을 빼기위해서 EnvironmentObject를 사용했습니다

gen-com replied:
설명 좋습니다. Published는 다시 정리해볼까요 ?

gen-com commented on week1/Memorize/Memorize/View/MemorizeButtonListVIew.swift:
```diff
@@ -0,0 +1,31 @@
+    @EnvironmentObject var gameManager: MemorizeCardGameManger
+    
+    var body: some View {
+        
```
공백 통일

gen-com commented on week1/Memorize/Memorize/View/MemorizeButtonListVIew.swift:
```diff
@@ -0,0 +1,31 @@
+            PlusMinusButtonVIew(mode: true)
+        }
+        .foregroundColor( gameManager.currentTheme.themeColor )
+        
```
무의미한 공백, 아래도 마찬가지

gen-com commented on week1/Memorize/Memorize/SubView/PlusMinusButtonVIew.swift:
```diff
@@ -0,0 +1,39 @@
+struct PlusMinusButtonVIew: View {
+    //true means plus mode
+    //false meas minus mode
+    var mode: Bool
```
mode를 나누는 것은 좋습니다. 다만 Bool이 아닌 enum으로 처리하면 더 명확하지 않을까요 ?

Dunkkkk replied:
Bool을 enum: ModeForPlusMinusButton으로 수정했습니다

gen-com commented on week1/Memorize/Memorize/SubView/PlusMinusButtonVIew.swift:
```diff
@@ -0,0 +1,39 @@
+
+import SwiftUI
+
+struct PlusMinusButtonVIew: View {
```
PlusMinusButtonV"I"ew -> PlusMinusButtonV"i"ew
Button과 View를 같이 쓰는 것은 중복입니다. Button은 View에 포함되니까요.

Dunkkkk replied:
관련 네이밍 전부 수정했습니다. 생각지도 못한건데 감사합니다. 이렇게 네이밍하는 습관들이겠습니다~

gen-com commented on week1/Memorize/Memorize/SubView/PlusMinusButtonVIew.swift:
```diff
@@ -0,0 +1,39 @@
+    //false meas minus mode
+    var mode: Bool
+    
+    @EnvironmentObject var gameManager: MemorizeCardGameManger
```
@EnvironmentObject ?

gen-com commented on week1/Memorize/Memorize/SubView/PlusMinusButtonVIew.swift:
```diff
@@ -0,0 +1,39 @@
+            Image(systemName: mode ? "plus.app" : "minus.square")
+                .font(.largeTitle)
+                .padding()
+          
```
무의미한 공백, 아래도 마찬가지

gen-com commented on week1/Memorize/Memorize/SubView/ThemeButtonListView.swift:
```diff
@@ -0,0 +1,47 @@
+struct ThemeButtonListView: View {
+    let buttonsImageAndTexts = [ImageAndText("globe.asia.australia", "Country"), ImageAndText("face.dashed","Emotion"), ImageAndText("fork.knife.circle","Food")]
+    
+    @EnvironmentObject var gameManager: MemorizeCardGameManger
```
@EnvironmentObject ?

gen-com commented on week1/Memorize/Memorize/SubView/ThemeButtonListView.swift:
```diff
@@ -0,0 +1,47 @@
+                ForEach(buttonsImageAndTexts, id: \.self){ content in
+                    Button(action: {}){
+                        ThemeButtonView(content: content)
+                            .onTapGesture {
```
button action을 사용하지 않고 onTapGesture 를 사용한 이유는 무엇인가요 ?

Dunkkkk replied:
수정완료했습니다~

gen-com commented on week1/Memorize/Memorize/SubView/ThemeButtonListView.swift:
```diff
@@ -0,0 +1,47 @@
+                    Button(action: {}){
+                        ThemeButtonView(content: content)
+                            .onTapGesture {
+                                switch content.Text {
```
switch 와 enum은 좋은 친구 입니다. 활용해 볼 수 있을까요 ?

gen-com commented on week1/Memorize/Memorize/SubView/ThemeButtonListView.swift:
```diff
@@ -0,0 +1,47 @@
+    @EnvironmentObject var gameManager: MemorizeCardGameManger
+    
+    var body: some View {
+        HStack{
```
중괄호 주의

HStack{ -> HStack {

Spacer를 양옆으로 뒀는데, alignmnet를 center로 하는 것과 무슨 차이가 있나요 ?

gen-com commented on week1/Memorize/Memorize/SubView/ThemeButtonView.swift:
```diff
@@ -0,0 +1,26 @@
+
+import SwiftUI
+
+struct ThemeButtonView: View{
```
중괄호 주의, 아래도 마찬가지

gen-com commented on week1/Memorize/Memorize/Singtone/MemorizeCardGameManger.swift:
```diff
@@ -0,0 +1,32 @@
+
+
+import SwiftUI
```
import SwiftUI 의 이유는 무엇인가요 ?

Dunkkkk replied:
깊게 생각하고 짠게 아닌 것 같습니다. 수정완료했습니다

gen-com commented on week1/Memorize/Memorize/Singtone/MemorizeCardGameManger.swift:
```diff
@@ -0,0 +1,32 @@
+import Foundation
+import Combine
+
+class MemorizeCardGameManger: ObservableObject{
```
중괄호 이슈 그리고 공백 통일

gen-com commented on week1/Memorize/Memorize/Singtone/MemorizeCardGameManger.swift:
```diff
@@ -0,0 +1,32 @@
+
+class MemorizeCardGameManger: ObservableObject{
+    
+    @Published var currentTheme: ThemeInfo = .country
```
@Published를 사용한 이유는 무엇인가요 ?

gen-com commented on week1/Memorize/Memorize/Singtone/MemorizeCardGameManger.swift:
```diff
@@ -0,0 +1,32 @@
+            currentCardNumber += 1
+        }
+    }
+
```
무의미한 공백

gen-com commented on week1/Memorize/Memorize/Singtone/MemorizeCardGameManger.swift:
```diff
@@ -0,0 +1,32 @@
+        }
+    }
+
+    func addOrSubtractCard(oneOrMinusOne num: Int) {
```
카드 수를 변경하는 함수를 통합해서 만들었습니다. 그 이유는 무엇인가요 ?

함수 이름과 outer parameter 이름까지 잘 읽히나요 ? 앞선 mode를 활용해보는 건 어떤가요 ?

Dunkkkk replied:
이 또한 모드의 rawVlaue를 통해 수정했습니다. outer parameter 이름도 수정했습니다

gen-com commented on week1/Memorize/Memorize/SubView/CardView.swift:
```diff
@@ -0,0 +1,49 @@
+
+import SwiftUI
+
+struct CardView: View, Memorize{
```
중괄호 주의

gen-com commented on week1/Memorize/Memorize/Singtone/MemorizeCardGameManger.swift:
```diff
@@ -0,0 +1,32 @@
+import SwiftUI
+import Foundation
+import Combine
+
```
폴더 이름이 싱글턴입니다. 이 클래스와 싱글턴이 어떤 관련이 있나요 ?

leejun6694 replied:
싱톤이야
근데 싱톤이 뭐야?

Dunkkkk replied:
처음 설계할 때, Static으로 구현했는데 @ObservableObject와 @EnvironmentObject를 발견하고 수정하는 과정에서 폴더명을 수정하지 못했습니다. 수정완료했습니다

gen-com commented on week1/Memorize/Memorize/Enum&Struct/MemorizeExtension.swift:
```diff
@@ -0,0 +1,33 @@
+    let Text: String
+    let distint: String
+    
+    init(_ Image: String , _ Text: String, _ distint: String = "0"){
```
outer pararmeter name을 _ 로 처리한 이유는 무엇인가요 ? 어떤 이점이 있나요 ?

Dunkkkk replied:
사용하는 범위가 적고, 기존에 튜플로 처리하던 것을 후에 hashable로 인해 struct로 만들면서 기존에 있던 코드를 고치지 않기 위해 _ 로 처리했습니다.

gen-com commented on week1/Memorize/Memorize/Enum&Struct/MemorizeExtension.swift:
```diff
@@ -0,0 +1,33 @@
+
+import SwiftUI
+
+protocol Memorize{
```
중괄호, 그리고 아무런 기능이 없는 프로토콜을 둔 이유는 무엇인가요 ?

Dunkkkk replied:
처음에 Memorize protocol로 공통적인 부분을 뺄려고 했는데 비효율적인 것 같아서 제거하였습니다

gen-com commented on week1/Memorize/Memorize/Enum&Struct/MemorizeExtension.swift:
```diff
@@ -0,0 +1,33 @@
+    
+}
+
+struct ImageAndText: Hashable, Identifiable{
```
ImageAndText, 이미지와 텍스트를 가지는 모든 것을 처리하는 건가요 ? 좀 더 역할에 따른 구체적인 네이밍은 없을까요 ?

그리고 중괄호 !

rose6649 replied:
ImageAndText를 묶어서 한 단위, 이름으로 더 잘보이게 만들면 좋을 것 같아요!

뒤에보면 Card에 쓰이는 것 같은데 그냥 Card 구조체를 만들어도 될 것 같네유

Dunkkkk replied:
네이밍 CardInfo로 수정하였습니다.

gen-com commented on week1/Memorize/Memorize/Enum&Struct/ThemeInfo.swift:
```diff
@@ -0,0 +1,49 @@
+import Foundation
+import SwiftUI
+
+enum ThemeInfo: Memorize {
```
: Memorize, 메모라이즈에 쓰인다를 표현하고 싶은 거면 Nested 형태로 구현하는 것은 어떤가요 ?

Dunkkkk replied:
Memorize는 후에 필요하지 않다고 필요하여 제거했습니다. 덕분에 Nested라는 것을 공부했습니다 감사합니다

gen-com commented on week1/Memorize/Memorize/Enum&Struct/MemorizeExtension.swift:
```diff
@@ -0,0 +1,33 @@
+//  Created by changgyo seo on 2022/07/25.
+//
+
+import SwiftUI
```
import SwiftUI ?

gen-com commented on week1/Memorize/Memorize/Enum&Struct/ThemeInfo.swift:
```diff
@@ -0,0 +1,49 @@
+//
+
+import Foundation
+import SwiftUI
```
import SwiftUI ?

gen-com commented on week1/Memorize/Memorize/Enum&Struct/ThemeInfo.swift:
```diff
@@ -0,0 +1,80 @@
+import SwiftUI
+
+
+enum ThemeInfo: String, Identifiable{
```
중괄호, 아래 공백 통일

gen-com commented on week1/Memorize/Memorize/Enum&Struct/ThemeInfo.swift:
```diff
@@ -0,0 +1,80 @@
+        }
+    }
+    
+    var maxCardCount: Int {
```
상수로 관리했을 때 유지 관리가 번거롭지 않을까요 ?

Dunkkkk replied:
이 부분을 수정하는 방법을 조금 해맸는데, 제가 생각하기에 멘토님들이 생각하는 좋은 바는 Associated value를 통해 입력받는 것이라고 생각해서 그 부분 수정했습니다.

gen-com commented on week1/Memorize/Memorize/Enum&Struct/ThemeInfo.swift:
```diff
@@ -0,0 +1,80 @@
+        }
+    }
+    
+    var powerOfImageList: [ImageAndText] {
```
powerOf 어떤 의미인가요 ?

Dunkkkk replied:
powerOf가 제곱이 아니라 두배가 된다고 착각했습니다. 죄송합니다 ㅜㅜ

leejun6694 commented on week1/Memorize/Memorize/Enum&Struct/MemorizeExtension.swift:
```diff
@@ -0,0 +1,26 @@
+    let Image: String
+    let Text: String
+    
+    //if user don't call third parameter, struct will be initalized to "0"
```
불필요한 주석은 제거해주세요

leejun6694 commented on week1/Memorize/Memorize/Enum&Struct/MemorizeExtension.swift:
```diff
@@ -0,0 +1,26 @@
+    
+    //use to be hashable this struct
+    var id: String
+    let Image: String
```
변수 이름은 소문자로 시작합니다.

leejun6694 commented on week1/Memorize/Memorize/Enum&Struct/ThemeInfo.swift:
```diff
@@ -0,0 +1,80 @@
+        }
+    }
+    
+    var themeRepresentativeImage: String {
```
Image 보단 ImageName 이 더 어울리겠네요~

leejun6694 commented on week1/Memorize/Memorize/Enum&Struct/MemorizeExtension.swift:
```diff
@@ -0,0 +1,26 @@
+    let Text: String
+    
+    //if user don't call third parameter, struct will be initalized to "0"
+    init(_ Image: String, _ Text: String, _ distinct: String = "0"){
```
중괄호 띄어쓰기

leejun6694 commented on week1/Memorize/Memorize/Enum&Struct/ThemeInfo.swift:
```diff
@@ -0,0 +1,80 @@
+    
+    var powerOfImageList: [ImageAndText] {
+        var list = imageList
+        imageList.forEach{ element in
```
생략 가능한 코드가 있어요~
참고 정도로만~~

Dunkkkk replied:
이 부분을 
return imageList + imageList
            .map { element -> CardInfo in
                return CardInfo(element.image, element.text, "1")
            }
이렇게 수정했는데, 생각하신 생략이 맞을까요?
이게 조금 더 읽기 좋은것 같긴해서 우선 수정후 push 했습니다~

leejun6694 replied:
$0
요런 것도 있어~~!

leejun6694 commented on week1/Memorize/Memorize/MemorizeApp.swift:
```diff
@@ -0,0 +1,21 @@
+//
+
+import SwiftUI
+
```
줄

leejun6694 commented on week1/Memorize/Memorize/MemorizeApp.swift:
```diff
@@ -0,0 +1,21 @@
+
+import SwiftUI
+
+
```
바

leejun6694 commented on week1/Memorize/Memorize/MemorizeApp.swift:
```diff
@@ -0,0 +1,21 @@
+import SwiftUI
+
+
+
```
꿈

leejun6694 commented on week1/Memorize/Memorize/Singtone/MemorizeCardGameManger.swift:
```diff
@@ -0,0 +1,46 @@
+    
+    func emitCard() -> [ImageAndText] {
+        var emitCardList = [ImageAndText]()
+        powerOfImageList.enumerated().forEach{ idx, element in
```
축약형이 아닌 의미 전달을 온전할 수 있게 풀네임으로 작성해주세요~
idx -> index

leejun6694 commented on week1/Memorize/Memorize/Singtone/MemorizeCardGameManger.swift:
```diff
@@ -0,0 +1,46 @@
+    func emitCard() -> [ImageAndText] {
+        var emitCardList = [ImageAndText]()
+        powerOfImageList.enumerated().forEach{ idx, element in
+            if( idx < currentCardCount){
```
띄어쓰기..!

leejun6694 replied:
swift 에선 if 조건문에 괄호를 보통 생략합니다

leejun6694 commented on week1/Memorize/Memorize/SubView/CardView.swift:
```diff
@@ -0,0 +1,42 @@
+            }
+        }
+        .onTapGesture {
+            isFaceUp.toggle()
```
어떤건 붙여서 쓰고, 어떤건 줄바꿈이 되어있고
규칙을 하나로 통일해주세요~

rose6649 commented on week1/Memorize/Memorize/Enum&Struct/ThemeInfo.swift:
```diff
@@ -0,0 +1,80 @@
+    case emotion = "Emotion"
+    case food = "Food"
+    
+    var id: String {
+        switch self{
```
띄어쓰기 주의보

rose6649 commented on week1/Memorize/Memorize/Enum&Struct/ThemeInfo.swift:
```diff
@@ -0,0 +1,80 @@
+        }
+    }
+    
+    var imageList: [ImageAndText] {
+        switch self{
```
주위보2

rose6649 commented on week1/Memorize/Memorize/Singtone/MemorizeCardGameManger.swift:
```diff
@@ -0,0 +1,46 @@
+
+class MemorizeCardGameManger: ObservableObject{
+    
+    static var themeList: [ThemeInfo] = [.country, .emotion, .food]
```
themeList는 enum의 성격을 가지고 있습니다.
ThemeInfo 는 지금 딱 세개의 case를 가지고 있네요, 그렇다면 CaseIterable 프로토콜을 한 번 찾아봐주세요 :)

Dunkkkk replied:
우선 CaseIterable같은 것을 찾고 있었는데, 있겠어? 하고 넘겼는데 감사합니다.
기본 프로토콜에 대해 조금 더 공부해야겠습니다 (ง •̀_•́)ง
하지만
위에 코드리뷰를 통해 수정하는 과정에서 Associated value를 사용하면  CaseIterable를 사용할 수 없다고 합니다.
공부해본바 compiler는 자동으로 해당 프로토콜에 맞는 적합성을 검사해 하나의 타입으로 자동합성한다고합니다.
근데 Associated value를 사용하면, 한 case에 여러 값이 합성이 될 수 있기 때문에 안된다고합니다.
그래서 연관값을 사용할까? CaseIterable를 사용할까? 하다가 우선 Associated value를 사용했습니다. ㅜㅜ 동시에 사용 할 수 있는 방법은 계속 공부해보겠습니다~

Dunkkkk committed:
Fix code convention and Change some enum

Fix code convention by code reviews and Change themeInfo by using associated value and all cases
and mode and isfaceup which is declared bool convert to enum

gen-com commented on week1/Memorize/Memorize/Enum&Struct/CommonEnum.swift:
```diff
@@ -0,0 +1,112 @@
+        }
+    }
+    
+    var imageList: [CardInfo] {
```
Computed로 두지 말고, 각 리스트를 static let으로 두면 어떤가요 ?
그러면 maximum count를 상수 말고 list.count로 할 수 있지 않을까요 ?

Dunkkkk replied:
뭔가 생각할때, 각 테마가 해당 카드의 리스트를 가지고 있는게 적합한거 같아서 enum에 computed로 했습니다.
static으로 하면, 뭐가 소유에 개념이 이상한거 같아서 안에 넣었습니다. ㅠㅠ
그리고 maximum count를 따로 둔것은 카드리스트보다 작은 수를 최대 카드수로 할 수도 있을 수 있으니 빼두었습니다. 영상을 잘못 파악한것 같습니다 수정하겠습니다!!

gen-com replied:
Computed면 imageList를 호출할 때마다 계산, 즉, CardInfo 배열을 매번 생성해서 보내는 건 인지하고 있나요 ?
imageList는 정적인 데이터인가요 동적인 데이터인가요 ?

Dunkkkk replied:
아! 감사합니다. 정적파일은 최대한 static let으로 관리하겠습니다.
그럼 혹시 멘토님은 정적인 저장프로퍼티를 관리할때, 혹시 어떤식으로 파일을 만드시는 지 궁금합니다. ㅜㅜ
따로 Static Property 같은 이름으로 구조체를 만들고 관리하시나요?

gen-com replied:
개인적으로 보통 상수값을 static let 처리를 하고, 관련된 구조체에 Nested로 따로 관리합니다.

최대한 보다는 상황에 맞게 잘 쓰는 것이 좋죠.

gen-com commented on week1/Memorize/Memorize/Enum&Struct/CommonEnum.swift:
```diff
@@ -0,0 +1,112 @@
+        }
+    }
+    
+    var doubleImageList: [CardInfo] {
```
```
var doubleImageList: [CardInfo] {
    imageList + imageList.map { CardInfo($0.image, $0.text, "1") }
}
```

Dunkkkk replied:
아!! 감사합니다.
혹시 $0, $1처럼 인자를 생략하는 것도 가독성이 좋은 건가요? 최대한 피할려고 노력했는데, 잘 모르겠습니다 ㅜㅜ

gen-com replied:
생략 가능하다는 것은 곧 충분히 유추할 수 있다는 사실이기도 하죠.
imageList에서 map을 호출합니다. 전달인자는 당연히 imageList의 원소겠죠 ?
반환하는 것을 생각해 봅시다. CardInfo를 반환한 다는 것을 추론할 수 있죠 ?(단일 문장에서는 return 생략 가능이니)

Dunkkkk replied:
넵! 감사합니다~~~

leejun6694 replied:
뭐야 여기서 했네 ㅋㅋㅋ

gen-com commented on week1/Memorize/Memorize/Enum&Struct/CommonEnum.swift:
```diff
@@ -0,0 +1,112 @@
+}
+
+enum ModeForPlusMinusButton: Int {
+    case plus = 1
```
그래서 공백 둘거에요 ?

gen-com commented on week1/Memorize/Memorize/Enum&Struct/CommonEnum.swift:
```diff
@@ -0,0 +1,112 @@
+}
+
+enum cardIsFaceUp {
+    case isFaceUp
```
안둘거에요 ?

Dunkkkk replied:
제가 리뷰를 잘못이해한것 같습니다. 혹시 어떤 공백 말씀하시는거세요? 죄송합니다.

gen-com commented on week1/Memorize/Memorize/Enum&Struct/CommonEnum.swift:
```diff
@@ -0,0 +1,112 @@
+import SwiftUI
+
+enum ThemeInfo: Identifiable {
+    
```
여기는 공백이 있는데요 ?

Dunkkkk committed:
Fix all Reviews

FIxed code convention about some enum type and use map opeeration using by contraction

leejun6694 commented:
고생했어요~~
아직 띄어쓰기가 잘 안되는 곳들이 있는데, 한번 더 봐줘~
